File: FrontEnd/lib/data/mappers/conversation_mapper.dart
/// @file conversation_mapper.dart
/// @description Implementación del mapper que transforma ConversationModel ↔ Conversation.
/// @module Chat
/// @layer Data
library;

import '../../domain/entities/conversation.dart';
import '../i_mappers/i_conversation_mapper.dart';
import '../models/conversation_model.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// CONVERSATION MAPPER
// ═══════════════════════════════════════════════════════════════════════════════

/// Convierte entre [ConversationModel] (serialización JSON) y [Conversation] (entidad de dominio).
///
/// [implements] IConversationMapper
/// [injectable] registrar en container.dart como IConversationMapper.
class ConversationMapper implements IConversationMapper {
  const ConversationMapper();

  /// Convierte un [ConversationModel] deserializado del JSON de backend en [Conversation].
  ///
  /// [param] model — modelo de serialización de la conversación.
  /// [returns] entidad de dominio limpia.
  @override
  Conversation toEntity(ConversationModel model) {
    return Conversation(
      id:                   model.id,
      participantId:        model.participantId,
      participantName:      model.participantName,
      participantPhoto:     model.participantPhoto,
      lastMessageText:      model.lastMessageText,
      lastMessageAt:        model.lastMessageAt != null
          ? DateTime.parse(model.lastMessageAt!).toUtc()
          : null,
      unreadCount:          model.unreadCount,
      createdAt:            DateTime.parse(model.createdAt).toUtc(),
      isParticipantDeleted: model.isParticipantDeleted,
    );
  }

  /// Convierte una [Conversation] de dominio en [ConversationModel] (para caché local).
  ///
  /// [param] entity — entidad de dominio de la conversación.
  /// [returns] modelo de serialización para Hive/SharedPreferences.
  @override
  ConversationModel toModel(Conversation entity) {
    return ConversationModel(
      id:                   entity.id,
      participantId:        entity.participantId,
      participantName:      entity.participantName,
      participantPhoto:     entity.participantPhoto,
      lastMessageText:      entity.lastMessageText,
      lastMessageAt:        entity.lastMessageAt?.toIso8601String(),
      unreadCount:          entity.unreadCount,
      createdAt:            entity.createdAt.toIso8601String(),
      isParticipantDeleted: entity.isParticipantDeleted,
    );
  }
}

File: BackEnd/src/data/mappers/conversation_mapper.ts
/**
 * @file conversation_mapper.ts
 * @description Implementación del mapper de conversaciones de chat.
 * Convierte entre ConversationDocument (MongoDB), Conversation (dominio) y ConversationResponseDTO (presentación).
 * @module Chat
 * @layer Data
 *
 * @implements {IConversationMapper}
 * @injectable
 */

import { injectable } from 'inversify';
import { ObjectId } from 'mongodb';
import type { IConversationMapper } from '../IMappers/IConversationMapper.js';
import { Conversation } from '../../domain/entities/Conversation.js';
import type { ConversationDocument } from '../datasources/mongodb/models/ConversationModel.js';
import type { ConversationResponseDTO, ParticipantSummaryDTO } from '../../domain/dtos/chat/conversation-response.dto.js';
import type { MessageResponseDTO } from '../../domain/dtos/chat/message-response.dto.js';

/**
 * Mapper de conversaciones de chat.
 *
 * @implements {IConversationMapper}
 * @injectable
 */
@injectable()
export class ConversationMapper implements IConversationMapper {

  /**
   * Convierte un documento MongoDB a entidad Conversation.
   *
   * @param doc — Documento de la colección 'conversations'.
   * @returns Entidad Conversation.
   */
  toEntity(doc: ConversationDocument): Conversation {
    return new Conversation({
      id:            doc._id.toHexString(),
      participants:  doc.participants.map(id => id.toHexString()),
      lastMessageAt: doc.lastMessageAt,
      createdAt:     doc.createdAt,
      updatedAt:     doc.updatedAt,
      deletedAt:     doc.deletedAt,
    });
  }

  /**
   * Construye un documento MongoDB para insertar una nueva conversación (sin _id).
   *
   * @param participantA — ID del primer participante.
   * @param participantB — ID del segundo participante.
   * @returns Documento parcial para insertar en MongoDB.
   */
  toDocument(participantA: string, participantB: string): Omit<ConversationDocument, '_id'> {
    const now = new Date();
    return {
      participants:  [new ObjectId(participantA), new ObjectId(participantB)],
      lastMessageAt: undefined,
      createdAt:     now,
      updatedAt:     now,
      deletedAt:     null,
    };
  }

  /**
   * Convierte una entidad Conversation al DTO de respuesta HTTP.
   *
   * @param entity — Entidad Conversation.
   * @param participant — Datos del otro participante.
   * @param lastMessage — Último mensaje de la conversación.
   * @param unreadCount — Número de mensajes no leídos.
   * @returns ConversationResponseDTO serializable.
   */
  toResponseDTO(
    entity: Conversation,
    participant: ParticipantSummaryDTO,
    lastMessage?: MessageResponseDTO,
    unreadCount = 0,
  ): ConversationResponseDTO {
    return {
      id:            entity.id,
      participant,
      lastMessage,
      lastMessageAt: entity.lastMessageAt?.toISOString(),
      unreadCount,
      createdAt:     entity.createdAt.toISOString(),
    };
  }
}
